### DISCUSSION TITLE
is there a way to resume on the calling thread?

### DISCUSSION THREAD

**Discussion by dr-matt** [David-Haim/concurrencpp#161] [Q&A]
Looking through the docs, this appears to be a very capable and well-thought-out library that wraps much of the lower-level c++ coroutine machinery. We have an application that requires certain operations to occur on the "main" thread of execution, and others to occur on a separate worker thread. In particular, we'd like to create a coroutine that

- does some work on the main/calling thread
- awaits the result of launching a worker task
- uses the result of that worker task to do more work on the original calling thread

Is `concurrencpp` able to accomplish this? Maybe something like the following:

```c++
concurrencpp::result<int> my_coro(concurrencpp::runtime& runtime) {
  std::cout << "first block is running on " << std::this_thread::get_id() << std::endl;
  // [do some work here on the calling thread]
  double first_result = 3.14159;

  std::string intermediate_result = co_await runtime.background_executor()->submit([&first_result]() {
    std::cout << "second block is running on " << std::this_thread::get_id() << std::endl;
    // [do intermediate work on the worker thread]
    return "hello world " + std::to_string(first_result);
  });

  co_await concurrencpp::resume_on(something);  // <--- possible to resume on the original thread??

  std::cout << "third block is running on " << std::this_thread::get_id() << std::endl;
  // [do final work here on the original calling thread]
  int final_result = intermediate_result.length();

  co_return final_result;
}
```

**Comment by David-Haim** [comment]
Hi there
Sorry for the late response. 

From what I understand, you want to resume on the main thread?
